File: python3/2375.py
class Solution:
    def smallestNumber(self, p: str) -> str:
        def find(res, p, trav, s, i, n):
            res[i] = chr(s + 48)
            if i == n:
                return True
            while True:
                s = s + (1 if (p[i] == "I") else -1)
                if s < 1 or s > 9:
                    return False
                if not trav[s]:
                    trav[s] = True
                    if find(res, p, trav, s, i + 1, n):
                        return True
                    trav[s] = False

        n = len(p)
        res = ["0"] * (n + 1)
        trav = [False] * 10
        for i in range(1, 10):
            trav[i] = True
            if find(res, p, trav, i, 0, n):
                return "".join(res)
            trav[i] = False
        return ""

File: cpp/2375.cpp
class Solution {
private:
    bool find(string& res, string& p, bool* trav, int s, int i, int n) {
        res[i] = (char)(s + '0');
        if (i == n)
            return true;
        while (true) {
            s = s + ((p[i] == 'I') ? 1 : -1);
            if (s < 1 || s > 9)
                return false;
            if (!trav[s]) {
                trav[s] = true;
                if (find(res, p, trav, s, i + 1, n))
                    return true;
                trav[s] = false;
            }
        }
    }

public:
    string smallestNumber(string& p) {
        int n = p.length();
        string res = p + "1";
        bool trav[10] = {false};
        for (int i = 1; i <= 9; i++) {
            trav[i] = true;
            if (find(res, p, trav, i, 0, n))
                return res;
            trav[i] = false;
        }
        return "";
    }
};
